utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,350 @@
+ * required functionality in that kernel. Hence, restrict usage of madvise to
+ * Linux and minherit to some BSD-flavored kernels. We do not consider Win.
+ *
+ * Android Trusty defines __linux__ for some reason, protect against that.
+ */
+#if defined(__linux__) && !defined(__TRUSTY__)
+#define USE_MADVISE
+#elif defined(__FreeBSD__) || defined(__OpenBSD__)
+#define USE_MINHERIT
+#endif
+
+#if defined(USE_MINHERIT) && defined(USE_MADVISE)
+#error "Both USE_MINHERIT and USE_MADVISE are defined. This should not be possible."
+#endif
+
+#if defined(USE_MADVISE) && defined(MADV_WIPEONFORK)
+S2N_STATIC_ASSERT(MADV_WIPEONFORK == 18);
+#else
+#define MADV_WIPEONFORK 18
+#endif
+
+/* These variables are used to disable all fork detection mechanisms or at the
+ * individual level during testing.
+ */
+#define S2N_FORK_DETECT_IGNORE 0
+#define S2N_FORK_DETECT_DO_NOT_IGNORE 1
+static int ignore_wipeonfork_or_inherit_zero_method_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+static int ignore_pthread_atfork_method_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+static int ignore_fork_detection_for_testing = S2N_FORK_DETECT_DO_NOT_IGNORE;
+
+#define FORK_EVENT 0
+#define NO_FORK_EVENT 1
+
+struct FGN_STATE {
+    /* The current cached fork generation number for this process */
+    uint64_t current_fork_generation_number;
+
+    /* Semaphore controlling access to the shared sentinel and signaling whether
+     * fork detection is enabled or not. We could use zero_on_fork_addr, but
+     * avoid overloading by using an explicit variable.
+     */
+    int is_fork_detection_enabled;
+
+    /* Sentinel that signals a fork event has occurred */
+    volatile char *zero_on_fork_addr;
+
+    pthread_once_t fork_detection_once;
+    pthread_rwlock_t fork_detection_rw_lock;
+};
+
+/* We only need a single statically initialised state. Note, the state is
+ * inherited by child processes.
+ */
+static struct FGN_STATE fgn_state = {
+    .current_fork_generation_number = 0,
+    .is_fork_detection_enabled = S2N_FAILURE,
+    .zero_on_fork_addr = NULL,
+    .fork_detection_once = PTHREAD_ONCE_INIT,
+    .fork_detection_rw_lock = PTHREAD_RWLOCK_INITIALIZER,
+};
+
+
+/* Can currently never fail. See initialise_fork_detection_methods() for
+ * motivation.
+ */
+static inline void initialise_wipeonfork_best_effort(void *addr, long page_size)
+{
+#if defined(USE_MADVISE)
+    /* Make sure that madvise() rejects invalid memory advice arguments.
+     * Some versions of qemu (up to at least 5.0.0-rc4, see
+     * linux-user/syscall.c) ignore madvise calls and just return zero (i.e.
+     * success). Therefore try an invalid call to check that the implementation
+     * of madvise is actually rejecting unknown "advice" values.
+     *
+     * This is not regarded as an error, but an inability to determine whether
+     * MADV_WIPEFORK is supported or not. We fall back to pthread_atfork if not.
+     */
+    if (madvise(addr, (size_t) page_size, -1) != 0) {
```

Comment:
Why do this check? We can't distinguish anyway. And correctness of `initialise_fork_detection_methods()` doesn't care about what actually happens here...